cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkPackageUtils.java

```diff
@@ -462,29 +466,63 @@ private static SparkFramework prepareSparkFramework(SparkCompat sparkCompat,
     String archiveName = "spark.archive-" + sparkVersion + "-" + VersionInfo.getVersion() + ".zip";
     Location frameworkDir = locationFactory.create("/framework/spark");
     Location frameworkLocation = frameworkDir.append(archiveName);
+    File archive = new File(tempDir, archiveName);
+    try (
+      ZipOutputStream zipOutput = new ZipOutputStream(new BufferedOutputStream(new FileOutputStream(archive)))) {
+      zipOutput.setLevel(Deflater.NO_COMPRESSION);
+      for (File file : getLocalSparkLibrary(sparkCompat)) {
+        zipOutput.putNextEntry(new ZipEntry(file.getName()));
+        Files.copy(file.toPath(), zipOutput);
+        zipOutput.closeEntry();
+      }
+    }
 
-    if (!frameworkLocation.exists()) {
-      File archive = new File(tempDir, archiveName);
-      try {
-        try (ZipOutputStream zipOutput = new ZipOutputStream(new BufferedOutputStream(new FileOutputStream(archive)))) {
-          zipOutput.setLevel(Deflater.NO_COMPRESSION);
-          for (File file : getLocalSparkLibrary(sparkCompat)) {
-            zipOutput.putNextEntry(new ZipEntry(file.getName()));
-            Files.copy(file.toPath(), zipOutput);
-            zipOutput.closeEntry();
+    Location writeInProgressLoc = frameworkDir.append("write_in_progress");
+    //Create writing file
+    boolean isWriteComplete = false;
+    while (!isWriteComplete) {
+      if (writeInProgressLoc.createNew()) {
+        Location tempDirHdfs = locationFactory.create("/framework/spark_temp_" + UUID.randomUUID());
+        Location tempArchiveHdfs = tempDirHdfs.append(archiveName);
+        tempDirHdfs.mkdirs("755");
+        try {
+          //Copy file to temp loc ( Upload )
+          try (OutputStream os = tempArchiveHdfs.getOutputStream()) {
+            Files.copy(archive.toPath(), os);
           }
-        }
-
-        // Upload spark archive to the framework location
-        frameworkDir.mkdirs("755");
 
-        try (OutputStream os = frameworkLocation.getOutputStream("644")) {
-          Files.copy(archive.toPath(), os);
+          //Move to actual location
+          tempArchiveHdfs.renameTo(frameworkLocation);
+          writeInProgressLoc.delete();
+          isWriteComplete = true;
+        } finally {
+          if (tempArchiveHdfs.exists()) {
+            tempArchiveHdfs.delete();
+          }
+          archive.delete();
+        }
+      } else {
```

Comment:
Please refactor the `if` and the `else` block into two private methods. This gives a better idea about what the intention is.